### ISSUE TITLE
Implement particle pusher by Higuera and Cary

### ISSUE THREAD

**Issue by MaxThevenet** [ECP-WarpX/WarpX#320]
It could be interesting to explore this particle pusher.
Here are the main references:
https://aip.scitation.org/doi/full/10.1063/1.4979989
http://accelconf.web.cern.ch/AccelConf/ipac2018/papers/thpml042.pdf
One could for instance mimic what is done for Boris and Vay pusher.

**Comment by jlvay** [comment]
We want to implement for sure the one described in the Phys. Plasmas paper. 
The conference paper describes a different pusher.

**Comment by RemiLehe** [comment]
@Yin-YinjianZhao 
Here is a bit more info on the implementation in WarpX:

- The particle pusher is called here:
https://github.com/ECP-WarpX/WarpX/blob/master/Source/Particles/PhysicalParticleContainer.cpp#L1486
https://github.com/ECP-WarpX/WarpX/blob/master/Source/Particles/PhysicalParticleContainer.cpp#L1593
https://github.com/ECP-WarpX/WarpX/blob/master/Source/Particles/RigidInjectedParticleContainer.cpp#L429
using the usual `ParallelFor` syntax which allows the code to work for both CPU and GPU.

- The definition of the Boris and Vay pusher (for single particle) are here:
https://github.com/ECP-WarpX/WarpX/blob/master/Source/Particles/Pusher/UpdateMomentumBoris.H
https://github.com/ECP-WarpX/WarpX/blob/master/Source/Particles/Pusher/UpdateMomentumVay.H
One would need to create a new file for the Higuera-Cary pusher.

- The pusher which is selected during a given simulation (i.e. Boris or Vay) depends on the user input. The code that does this is here:
https://github.com/ECP-WarpX/WarpX/blob/master/Source/Utils/WarpXAlgorithmSelection.cpp
https://github.com/ECP-WarpX/WarpX/blob/master/Source/Utils/WarpXAlgorithmSelection.H
This would need to be updated.

- Finally, the documentation would need to be updated here:
https://github.com/ECP-WarpX/WarpX/blob/master/Docs/source/running_cpp/parameters.rst
(Search for `algo.particle_pusher` in that file.)

**Comment by RemiLehe** [comment]
@jlvay also pointed out a recent article that compares the different solvers:
https://iopscience.iop.org/article/10.3847/1538-4365/aab114/meta

**Comment by Yin-YinjianZhao** [comment]
When reading the papers and implementing the methods, I found two typos in two papers, which made me suffer for some time before figuring them out...
So, I'd like to share them here. See the two photos below.

For testing, I followed Ripperda's paper and tested Boris's, Vay's, and Higuera's methods using five cases (uniform E, uniform B, force-free, E x B, magnetic mirror). Although I couldn't get exactly the same results as those shown in Ripperda's paper, my results are good enough to indicate that my implementation is correct.
The next step will be to add the Higuera's solver into WarpX.
And if we want to add some automated tests for WarpX for the pushers, the ones in Ripperda's paper are good candidates.

![image](https://user-images.githubusercontent.com/56095356/66685475-83c83700-ec31-11e9-9e84-b17cc67dca7b.png)

![image](https://user-images.githubusercontent.com/56095356/66685537-b8d48980-ec31-11e9-95a9-dfd0c4cf3c89.png)
